mindsdb/integrations/handlers/huggingface_handler/huggingface_handler.py

```diff
@@ -346,38 +347,131 @@ def describe(self, attribute: Optional[str] = None) -> pd.DataFrame:
+        labels = finetune_args.get("labels")
+        if not labels:
+            raise ValueError("Labels not specified in arguments")
+        
+        # Create labels_map if it doesn't exist
+        if "labels_map" not in finetune_args:
+            finetune_args["labels_map"] = {label: i for i, label in enumerate(labels)}
+        
+        label_map = finetune_args["labels_map"]
+        df["labels"] = df["labels"].map(label_map)
+
+        print("Finetune args after processing:", finetune_args)
+        print("DataFrame head:", df.head())
+        print("DataFrame info:", df.info())
+
+        tokenizer, trainer = FINETUNE_MAP[task](df, finetune_args)
 
         try:
-            trainer.train()
-            trainer.save_model(
-                model_folder
-            )  # TODO: save entire pipeline instead https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.Pipeline.save_pretrained
+            # Checks Transformers version
+            transformers_version = version.parse(transformers.__version__)
+            print(f"Transformers version: {transformers_version}")
+
+            if transformers_version >= version.parse("4.0.0"):
+                # Set up early stopping for newer versions
+                early_stopping = transformers.EarlyStoppingCallback(early_stopping_patience=3)
+                try:
+                    trainer.train(callbacks=[early_stopping])
+                except TypeError:
+                    print("Warning: Callbacks not supported in this version of Transformers. Training without early stopping.")
+                    trainer.train()
+            else:
+                # For older versions, train without callbacks
+                print("Warning: Using an older version of Transformers. Early stopping is not available.")
+                trainer.train()
+            
+            eval_results = trainer.evaluate()
+            print(f"Evaluation results: {eval_results}")
+
+            trainer.save_model(model_folder) 
+            # TODO: save entire pipeline instead https://huggingface.co/docs/transformers/main_classes/
+            # pipelines#transformers.Pipeline.save_pretrained
             tokenizer.save_pretrained(model_folder)
 
+            finetune_args["fine_tuned"] = True
+            finetune_args["eval_results"] = eval_results
             # persist changes
-            self.model_storage.json_set("args", args)
+            self.model_storage.json_set("args", finetune_args)
             self.model_storage.folder_sync(model_folder_name)
 
+            print("Fine-tuning completed successfully")
+            print("Final evaluation results:", eval_results)
+
         except Exception as e:
             err_str = f"Finetune failed with error: {str(e)}"
-            logger.debug(err_str)
+            print(err_str)
+            logger.error(err_str)
             raise Exception(err_str)
+    
+    def print_debug_info(self):
```

Comment:
To improve maintainability and allow better control over log levels, could you please replace the print statements with a logger? Also instead of placing a logger call on every line, feel free to group related log messages where appropriate to avoid excessive logging.

Reply:
@ZoranPandovski  thank you for reviewing this PR
I will make the changes and update the PR